File: wakakusa/demo/charts.rb
# Charts that can say what they mean: a profit-and-loss bar chart whose
# losing months hang below the zero line, and a two-series line chart.
# `min`/`max` both zero take the range from the data; `axis` draws the
# tick labels and a faint gridline at each; `series` takes one list per
# line, `colors` one color each.
require "wakakusa"

HEADING = { size: 18.0, color: "accent" }.freeze
FAINT = { size: 12.0, color: "#8a8f98" }.freeze

class Book
  def initialize
    @months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun"]
    @profit = [12.0, -8.0, 4.0, -3.0, 15.0, -6.0]
    @requests = [40.0, 55.0, 48.0, 62.0, 70.0, 58.0]
    @errors = [3.0, 9.0, 5.0, 12.0, 6.0, 4.0]
    @traffic = [@requests, @errors]
    @n = 6
  end

  def next_month
    @n += 1
    # A deterministic next month, so both runs read the same numbers
    # and the gate can compare them.
    @profit = grown(@profit, (@n * 7 % 41).to_f - 18.0)
    @months = grown(@months, "M#{@n}")
    @requests = grown(@requests, (@n * 13 % 50).to_f + 30.0)
    @errors = grown(@errors, (@n * 5 % 14).to_f)
    @traffic = [@requests, @errors]
  end

  # `list + [one]` on a field takes a different shape in the two runs;
  # a copy that is pushed to takes the same one in both.
  def grown(list, item)
    out = list.dup
    out.push(item)
    out
  end

  def view
    column(
      text("Profit and loss", **HEADING),
      text("negative months hang below the zero line", **FAINT),
      bar_chart(@profit, labels: @months, axis: true, height: 150.0),
      text("Traffic", **HEADING),
      text("requests and errors, one color each", **FAINT),
      line_chart(series: @traffic, labels: @months, colors: ["accent", "#f38ba8"],
                 axis: true, max: 90.0, height: 150.0),
      row(button("next month") { next_month }, spacing: 8.0),
      spacing: 12.0,
      padding: 16.0
    )
  end
end

run(Book.new, title: "charts")

File: gomamochi/demo/charts.go
// Charts that can say what they mean: a profit-and-loss bar chart whose
// losing months hang below the zero line, and a two-series line chart.
// `Min`/`Max` both zero take the range from the data; `Axis` draws the
// tick labels and a faint gridline at each; `Series` takes one list per
// line, `Colors` one color each.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Book struct {
	months   []string
	profit   []float64
	requests []float64
	errors   []float64
	n        int
}

func (b *Book) nextMonth() {
	b.n += 1
	// A deterministic next month, so both runs read the same numbers
	// and the gate can compare them.
	b.profit = append(b.profit, float64(b.n*7%41)-18)
	b.months = append(b.months, fmt.Sprintf("M%d", b.n))
	b.requests = append(b.requests, float64(b.n*13%50)+30)
	b.errors = append(b.errors, float64(b.n*5%14))
}

func (b *Book) View() Element {
	return Column(
		Text("Profit and loss").Size(18).Color("accent"),
		Text("negative months hang below the zero line").Size(12).Color("#8a8f98"),
		BarChart(b.profit).Labels(b.months...).Axis(true).Height(150),
		Text("Traffic").Size(18).Color("accent"),
		Text("requests and errors, one color each").Size(12).Color("#8a8f98"),
		LineChart(nil).Series([][]float64{b.requests, b.errors}).Labels(b.months...).
			Colors("accent", "#f38ba8").Axis(true).Max(90).Height(150),
		Row(Button("next month").OnClick(func() { b.nextMonth() })).Spacing(8),
	).Spacing(12).Padding(16)
}

func main() {
	Run(&Book{
		months:   []string{"Jan", "Feb", "Mar", "Apr", "May", "Jun"},
		profit:   []float64{12, -8, 4, -3, 15, -6},
		requests: []float64{40, 55, 48, 62, 70, 58},
		errors:   []float64{3, 9, 5, 12, 6, 4},
		n:        6,
	}, Title("charts"))
}
